### PR TITLE
Assignment 8

### PR DESCRIPTION
**Pull request by dnisyndrum**

### CODE REVIEW

File: Assignment/Person.cs [hunk 1]
```diff
@@ -8,7 +10,21 @@ public class Person : IPerson
         public string City { get; set; }
         public string State { get; set; }
         public string Zip { get; set; }
+        public string Email { get; set; }
 
         public Address Address { get;set; }
+
+        public Person(string fname, string lname, string streetAddress, string city, string state, string zip, string email)
```
Comment (jacob-berger) [inline Assignment/Person.cs]: Creating your person this way, you end up with Person.Address.(City/State/Zip/Street) as well as a reference from Person.(City/State/Zip/Street).

File: Assignment/SampleData.cs [hunk 2]
```diff
@@ -1,30 +1,70 @@
...
-            => throw new NotImplementedException();
+        private string _FilePath;
 
-        // 3.
-        public string GetAggregateSortedListOfStatesUsingCsvRows()
-            => throw new NotImplementedException();
+        enum Columns
+        {
+            Id, FirstName, LastName, Email, StreetAddress, City, State, Zip
+        }
```
Comment (jacob-berger) [inline Assignment/SampleData.cs]: Good idea to label your columns.

File: Assignment/SampleData.cs [hunk 3]
```diff
@@ -1,30 +1,70 @@
...
+            {
+                var people = from line in CsvRows select CreatePerson(line);
+                return people;
+            }
+        }
 
-        // 5.
-        public IEnumerable<(string FirstName, string LastName)> FilterByEmailAddress(
-            Predicate<string> filter) => throw new NotImplementedException();
+        public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()
```
Comment (jacob-berger) [inline Assignment/SampleData.cs]: Should be sorting first by state, then city, then zip.

**Comment by MarkMichaelis** [comment]
## { FullName } ({GitHubId})

## {BaseBranch}

| Build/Compile ({BuildResultsSummary})| **10** |
| --------------------- | ------ |
| Assignment compiles even though major pieces might be missing, or there are significant problems: **{BuildSuccessfulCheck}** |  **{BuildCompilesPoints}** |
| The assignment contains only a few required elements or has significant problems. | **5**


| Assignment  | **17** |
| --------------------- | ------ |
| 1. Implement the `ISampleData.CsvRows` property, loading the data from the `People.csv` file and returning each line as a single string. ✔ <br/>- Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔ <br/> - Using LINQ, skip the first row in the `People.csv`. ✔ <br/> - Be sure to appropriately handle resource (`IDisposable`) items correctly if applicable (and it may not be depending on how you implement it). ✔  | ✔ |
| 2. Implement `IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()` to return a **sorted**, **unique** list of states. ✔ <br/> Use `ISampleData.CsvRows` for your data source. ✔ <br/>- Don't forget the list should be unique. ✔ <br/>- Sort the list alphabetically ✔ <br/>- Include a test that leverages a hard coded list of Spokane based addresses. ❌ <br/>- Include a test that uses LINQ to verify the data is sorted correctly (do not use a hard coded list). ✔ | ✔ |
| 3. Implement `ISampleData.GetAggregateSortedListOfStatesUsingCsvRows()` to return a `string` that contains a **unique**, comma separated list of states. ✔ <br/>- Use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows` for your data source. ✔ <br/>- Given the array, consider using `string.Join` to combine the list into a single string. | ✔ |
| 4.Implement the `ISampleData.People` property to return all the items in `People.csv` as `Person` objects. ✔ <br/>- Use `ISampleData.CsvRows` as the source of the data. ✔ <br/>- Sort the list by State, City, Zip. (Sort the addresses first then select). ❌ <br/>- Be sure that `Person.Address` is also populated. <br/>- Adding null validation to all the `Person` and `Address` properties is **optional**. <br/>- Consider using `ISampleData.CsvRows` in your test to verify your results. | ✔ |
| 5. Implement `ISampleDate.FilterByEmailAddress(Predicate<string> filter)` to return a list of names where the email address matches the `filter`. ✔ <br/>- Use `ISampleData.People` for your data source. ✔ | ✔ |
| 6. Implement `ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people)` to return a `string` that contains a **unique**, comma separated list of states. ✔ <br/>- Use the `people` parameter for your data source. ✔ <br/>- At a minimum, use `System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔ <br/>- Don't forget the list should be unique. ✔ <br/>- It is recommended that at a minimum you use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows` to validate your result. | ✔ |


| Project Analysis |   |
| -----------------| - |
| Nullable Enabled: | **✔** |
| CodeReview Performed | **✔** |
| Appropriately handle all warnings | **✔** |
| Add the code analysis packages* |  **✔** |
*<sub><sup>`IntelliTectAnalyzer.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.CodeQuality.Analyzers.dll`,`Microsoft.NetCore.Analyzers.dll`,`Microsoft.NetFramework.Analyzers.dll`. (It is okay to appropriately disable rules such as  localization/culture rules.)</sub></sup>


{ ProjectDetails }

| Testing | **20** |
| --------------------- | ------ |
| {TestResultsSummary} | **90%** |
| Unit test project fails to compile, most test fails, or contains significant problems. | **10** |
| Most assigned work is tested or contain some errors. | **5** |
| All assigned work is tested | **5** |


| Extra Credit | **0** |
| --------------------- | ------ |
| Write awesome, well tested code that follows all the guidelines outlined in the book. | ❌ |
| Add support for an `IReadOnlyArray` interface (for you to define) that provides readonly access to the count and each element in the array.  Be sure to appropriately decorate it for covariance and contravariance. | ❌ |
| Implement all homework using async/await and multi-threading by defining a new `SampleDataAsync` class that implements `ISampleData`). Refactor both your SampleData and SampleDataAsync classes so there is minimal duplication.  This includes refactoring your tests so a significant amount of test code can be re-used.| ❌ |

**Pull request closed without merging**